qiskit_optimization/translators/model_translator.py

```diff
@@ -0,0 +1,64 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Abstract class for optimization model translators"""
+
+from abc import ABC, abstractmethod
+from typing import Any, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    # pylint: disable=cyclic-import
+    from qiskit_optimization.problems.quadratic_program import QuadraticProgram
+
+
+class ModelTranslator(ABC):
+    """Translator between an optimization model and a quadratic program
+
+    Note:
+        The types of quadratic_program is `Any` because we need to avoid cyclic import.
+    """
+
+    @abstractmethod
+    def is_compatible(self, model: Any) -> bool:
```

Comment:
In the docplex case we have that installed always right as a dependent. If I pass a translator in a list, I am looking at code below that goes over the list checking is_compatible and is so translates it, if the necessary optimization model code is part of an optional install, what do we expect to happen. Should the translator fail to load and throw an exception as the dependent is not installed - or do we expect the user to pass in the list as some sort of preferred order of translation whereby we pick the one that works i.e the translator will still load but say its not compatible, not because its cannot truly do it, but because the optional dependence it needs is not installed and thus is not presently compatible until such time, if ever, that were changed.

Reply:
If users try to translate an optimization model of a dependent library, the model is made by the library. So the library is supposed to be installed in their environment and the translator can catch the model. However, it might be convenient to have a method `is_installed` to check whether the library is installed or not.